File: packages/ios/OpenCode/Views/ShellStore.swift
import SwiftUI

/// One command + its output in the pocket terminal.
struct ShellEntry: Identifiable {
    let id = UUID()
    let command: String
    let output: String
}

/// Keeps each session's shell history alive across opening/closing the terminal
/// sheet (and navigating away and back), so tapping Done no longer throws away
/// what you ran. In-memory for the app's lifetime, keyed by session id.
///
/// NOTE: this is the one-shot `/session/:id/shell` model (each command runs
/// independently — no shared cwd/env). Real persistent, interactive, multi-tab
/// terminals are a follow-up on the server's PTY API.
@Observable
@MainActor
final class ShellStore {
    static let shared = ShellStore()
    private init() {}

    private var histories: [String: [ShellEntry]] = [:]

    func history(_ sessionID: String) -> [ShellEntry] { histories[sessionID] ?? [] }
    func append(_ entry: ShellEntry, to sessionID: String) {
        histories[sessionID, default: []].append(entry)
    }
    func clear(_ sessionID: String) { histories[sessionID] = [] }
}

File: packages/android/app/src/main/java/studio/eugenezakharov/opencode/ui/session/ShellStore.kt
package studio.eugenezakharov.opencode.ui.session

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.snapshots.SnapshotStateList

/**
 * Keeps each session's shell history alive across opening/closing the terminal
 * dialog (and navigating away and back), so Close no longer discards what you
 * ran. In-memory for the process lifetime, keyed by session id. Compose reads it
 * reactively. Mirrors the iOS ShellStore.
 *
 * NOTE: still the one-shot `/session/:id/shell` model (each command independent).
 * Real persistent, interactive, multi-tab terminals are a follow-up on the PTY API.
 */
object ShellStore {
    private val histories = mutableStateMapOf<String, SnapshotStateList<Pair<String, String>>>()

    /** A session's command+output history, created on first access. */
    fun history(sessionId: String): SnapshotStateList<Pair<String, String>> =
        histories.getOrPut(sessionId) { mutableStateListOf() }

    fun clear(sessionId: String) {
        histories[sessionId]?.clear()
    }
}
